sips/pending/_posts/2017-02-07-make-types-behave-like-expressions.md

```diff
@@ -0,0 +1,197 @@
+| Feb 7th 2017  | Initial Draft    |
+
+---
+## Introduction
+Currently scala allows symbol operators (`-`, `*`, `~~>`, etc.) for both type names and definition names.
+Unfortunately, there is a 'surprise' element since the two differ in behaviour:
+
+* **Infix operator precedence and associativity**: 
+Infix types are 'mostly' left-associative, 
+while the expression operations are more intuitive with different precedence weights.
+Please see [Infix Types](http://scala-lang.org/files/archive/spec/2.12/03-types.html#infix-types) and [Infix Operations](http://scala-lang.org/files/archive/spec/2.12/06-expressions.html#infix-operations) sections of the Scala specifications for more details. 
+
+**Example**:
+```scala
+object InfixExpressionPrecedence {
+    case class Nummy(expand : String) {
+      def + (that : Nummy) : Nummy = Nummy(s"Plus[$this,$that]")
+      def * (that : Nummy) : Nummy = Nummy(s"Prod[$this,$that]")
+      override def toString : String = expand
+    }
+    object N1 extends Nummy("N1")
+    object N2 extends Nummy("N2")
+    object N3 extends Nummy("N3")
+    object N4 extends Nummy("N4")
+    val result_expected = N1 + N2 * N3 + N4
+    //result_expected.expand is Plus[Plus[N1,Prod[N2,N3]],N4]
+}
+object InfixTypePrecedence {
+    trait Plus[N1, N2]
+    trait Prod[N1, N2]
+    type +[N1, N2] = Plus[N1, N2]
+    type *[N1, N2] = Prod[N1, N2]
+    trait N1 
+    trait N2
+    trait N3 
+    trait N4 
+    type Result_Surprise = N1 + N2 * N3 + N4
+    //Result_Surprise expands to Plus[Prod[Plus[N1,N2],N3],N4]
+    type Result_Expected = N1 + (N2 * N3) + N4
+    //Result_Expected expands to Plus[Plus[N1,Prod[N2,N3]],N4]
+}
+```
+
+* **Prefix operators bracketless unary use**: While expressions have prefix unary operators, there are none for types. See the [Prefix Operations](http://scala-lang.org/files/archive/spec/2.12/06-expressions.html#prefix-operations) section of the Scala specification. 
+This is a lacking feature of the type language Scala offers. See also interactions of this feature with other Scala features, further down this text. 
+(Author's note: Not crucial as infix precedence, but good for completeness) 
+
+**Example**:
+```scala
+object PrefixExpression {
+  case class Nummy(expand : String) {
+    def unary_- : Nummy = Nummy(s"-$this")
+    def unary_~ : Nummy = Nummy(s"~$this")
+    def unary_! : Nummy = Nummy(s"!$this")
+    def unary_+ : Nummy = Nummy(s"+$this")
+  }
+  object N extends Nummy("N")
+  val n1 = -N
+  val n2 = ~N
+  val n3 = !N
+  val n4 = +N
+}
+object NonExistingPrefixTypes {
+  trait unary_-[A]
+  trait unary_~[A]
+  trait unary_![A]
+  trait unary_+[A]
+  trait N
+  type N1 = -N //Not working
+  type N2 = ~N //Not working
+  type N3 = !N //Not working
+  type N4 = +N //Not working
+}
+```
+
+---
+## Proposal
+The proposal is split into two; type infix precedence, and prefix unary types. Note to the SIP committee: It might be better to vote on the two parts separately.   
```

Comment:
Let's do it this way, yes. We'll see how this experiment goes, we haven't actually had this situation happened yet.